File: 011/main.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let len = height.length
  let i = 0
  let j = len - 1
  let max = compute(i, j, height[i], height[j])
  let s
  while (i < j) {
    let hi = height[i]
    let hj = height[j]
    if (hi < hj) {
      while (hi < hj && i < j) {
        i++
        hi = height[i]
        s = compute(i, j, hi, hj)
        if (s > max) max = s
      }
    } else {
      while (hi >= hj && i < j) {
        j--
        hj = height[j]
        s = compute(i, j, hi, hj)
        if (s > max) max = s
      }
    }
  }
  return max
}

function compute (i, j, hi, hj) {
  return (j - i) * Math.min(hi, hj)
}

console.log(maxArea([1, 4, 7, 9, 11, 12]))
console.log(maxArea([1, 1]))

File: 011/main.cpp
#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int len = height.size();
        int i = 0;
        int j = len - 1;
        int ans = (j - i) * min(height[i], height[j]);
        int tmp;
        while (i < j) {
            int hi = height[i];
            int hj = height[j];
            if (hi < hj) {
                while (hi < hj && i < j) {
                    i++;
                    hi = height[i];
                    tmp = (j - i) * min(hi, hj);
                    if (tmp > ans) ans = tmp;
                }
            } else {
                while (hi >= hj && i < j) {
                    j--;
                    hj = height[j];
                    tmp = (j - i) * min(hi, hj);
                    if (tmp > ans) ans = tmp;
                }
            }
        }
        return ans;
    }
};

int main()
{
    Solution sol;
    vector<int> height;
    height.push_back(1);
    height.push_back(2);
    height.push_back(3);
    height.push_back(4);
    int res = sol.maxArea(height);
    cout << res << endl;
    return 0;
}
